### PR TITLE
Disk-based favorites cache 

### PR DESCRIPTION
**Pull request by aaazalea**
Still needs
 [x] favorites cache merged
 [ ] client PR merged
 [ ] more tests

### CODE REVIEW

**Comment by jakob223** [comment]
Client PR: https://github.com/keybase/client/pull/14608

**Comment by jakob223** [comment]
Expect CircleCI here to fail until client is merged. I definitely think this should get a `kbfsdocker` test or two, not sure if any tests in this repo would be useful.

File: libkbfs/favorites.go [hunk 1]
```diff
@@ -107,6 +112,87 @@ func NewFavorites(config Config) *Favorites {
 	return newFavoritesWithChan(config, make(chan *favReq, 100))
 }
 
+func (f *Favorites) readCacheFromDisk(ctx context.Context) error {
+	// Read the encrypted cache from disk
+	folder := f.config.StorageRoot()
+	subfolder := "kbfs.favorites.cache"
```
Comment (strib) [inline libkbfs/favorites.go]: This is screaming for a global constant.

File: libkbfs/favorites.go [hunk 2]
```diff
@@ -107,6 +112,87 @@ func NewFavorites(config Config) *Favorites {
...
+	}
+
+	// Send the data to the service to be decrypted
+	decryptedData, err := f.config.KeybaseService().DecryptFavorites(ctx, data)
+	if err != nil {
+		return err
+	}
+
+	// Decode the data into the a map
+	decoder := gob.NewDecoder(bytes.NewBuffer(decryptedData))
```
Comment (strib) [inline libkbfs/favorites.go]: Why use gob?  As far as I know this is the only place we use this.

I think a more standard way would be to define a struct that holds the cache map, and encode that using `f.config.Codec()` (which would use msgpack). Besides being more standard, that would let us add other fields in the future.

Also it might be good to add a version number to the struct holding the map as well, in case the format (or interpretation of the format) of the cache needs to change in the future.  And after decrypting, we check that the version number matches what we expect or else we panic.

File: libkbfs/favorites.go [hunk 3]
```diff
@@ -107,6 +112,87 @@ func NewFavorites(config Config) *Favorites {
...
+	folder := f.config.StorageRoot()
+	subfolder := "kbfs.favorites.cache"
+	session, err := f.config.KBPKI().GetCurrentSession(ctx)
+	if err != nil {
+		return err
+	}
+	user := string(session.UID)
+	path := filepath.Join(folder, subfolder, user)
+	// TODO: is this the right permissions
+	return ioutil.WriteFile(path, data, 0600)
```
Comment (strib) [inline libkbfs/favorites.go]: Besides using msgpack for the encoded data we send over to the service (as mentioned above), I think we probably want to also use it here for the contents of the file on disk (i.e., a struct that contains a `[]byte`).  That's because we should also add a version number to that struct also, in case we ever change the encryption scheme of the cache in the future.  And when reading the file, check the version number and panic if it's unexpected.

File: libkbfs/favorites.go [hunk 4]
```diff
@@ -107,6 +112,87 @@ func NewFavorites(config Config) *Favorites {
...
+	go f.loop()
+}
+
+// Initialize loads the favorites cache from disk and starts listening for
+// requests asynchronously.
+func (f *Favorites) Initialize(ctx context.Context) {
+	// load cache from disk
+	err := f.readCacheFromDisk(ctx)
+	if err != nil {
+		f.config.MakeLogger("favorites").CDebugf(nil,
```
Comment (strib) [inline libkbfs/favorites.go]: Maybe `CWarningf` instead?

File: libkbfs/favorites.go [hunk 5]
```diff
@@ -125,15 +211,18 @@ func (f *Favorites) handleReq(req *favReq) (err error) {
 	//  * The user asked us to refresh
 	//  * We haven't fetched it before
 	//  * It's stale
-	if req.refresh || f.cache == nil || time.Now().After(f.cacheExpireTime) {
+	if req.refresh || f.cache == nil || time.Now().After(f.
```
Comment (strib) [inline libkbfs/favorites.go]: Please split the line after the last `(`.

Comment (strib) [inline libkbfs/favorites.go]: Please use `f.config.Clock().Now()` instead of `time.Now()`.  (I missed that in an unreviewed update to an old PR, sorry.)

File: libkbfs/favorites.go [hunk 6]
```diff
@@ -125,15 +211,18 @@ func (f *Favorites) handleReq(req *favReq) (err error) {
...
 		folders, err := kbpki.FavoriteList(req.ctx)
 		if err != nil {
 			if req.refresh {
 				// if we're supposed to refresh the cache and it's not
-				// working, try again in a few minutes.
-				in5Minutes := time.Now().Add(5 * time.Minute)
-				if in5Minutes.After(f.cacheExpireTime) {
-					f.cacheExpireTime = in5Minutes
+				// working, try again in a minute.
+				in1Minute := time.Now().Add(time.Minute)
```
Comment (strib) [inline libkbfs/favorites.go]: Why the change here?  Mostly just curious.

File: libkbfs/favorites.go (hunk 4 above)
Comment (jakob223) [reply libkbfs/favorites.go]: We do expect this to happen on the first login of each user; does that qualify for a warning?

File: libkbfs/favorites.go (hunk 6 above)
Comment (jakob223) [reply libkbfs/favorites.go]: I don't remember. I think one minute seems more reasonable to me if the reason for failure is that the computer is starting up and network doesn't exist yet.

File: libkbfs/favorites.go (hunk 4 above)
Comment (strib) [reply libkbfs/favorites.go]: You mean we expect it to happen _before_ the first login of each user?  In that case, debug is fine.

File: libkbfs/favorites.go [hunk 7]
```diff
@@ -107,6 +112,87 @@ func NewFavorites(config Config) *Favorites {
...
+
+	// Write the encrypted cache to disk
+	folder := f.config.StorageRoot()
+	subfolder := "kbfs.favorites.cache"
+	session, err := f.config.KBPKI().GetCurrentSession(ctx)
+	if err != nil {
+		return err
+	}
+	user := string(session.UID)
+	path := filepath.Join(folder, subfolder, user)
```
Comment (strib) [inline libkbfs/favorites.go]: Just realized that this path has to have the UID in there somewhere.  Multiple Keybase accounts should be able to cache their favorites to disk.

Comment (strib) [reply libkbfs/favorites.go]: Nevermind, I'm dumb.

File: libkbfs/favorites.go [hunk 8]
```diff
@@ -107,6 +113,118 @@ func NewFavorites(config Config) *Favorites {
...
+	}
+
+	// decode the data from the file and ensure its version is correct
+	var decodedData favoritesCacheEncryptedForDisk
+	err = f.config.Codec().Decode(data, &decodedData)
+	if err != nil {
+		return err
+	}
+	if decodedData.version != favoritesDiskCacheStorageVersion {
+		panic("Favorites cache storage version is wrong.")
```
Comment (strib) [inline libkbfs/favorites.go]: I think I changed my mind about panicking here (and below).  Just return an error.

Oh, and you might as well include the unexpected version number in the error string.

File: libkbfs/favorites.go [hunk 9]
```diff
@@ -125,15 +243,18 @@ func (f *Favorites) handleReq(req *favReq) (err error) {
...
 		folders, err := kbpki.FavoriteList(req.ctx)
 		if err != nil {
 			if req.refresh {
 				// if we're supposed to refresh the cache and it's not
-				// working, try again in a few minutes.
-				in5Minutes := time.Now().Add(5 * time.Minute)
-				if in5Minutes.After(f.cacheExpireTime) {
-					f.cacheExpireTime = in5Minutes
+				// working, try again in a minute.
+				in1Minute := f.config.Clock().Now().Add(time.Minute)
```
Comment (strib) [inline libkbfs/favorites.go]: Please also update the `time.Now()` in the `ForceWallClock` call below.
